rest-api-model-ql/src/main/kotlin/org/modelix/sample/restapimodelql/LightModelClientWrapper.kt

```diff
@@ -0,0 +1,120 @@
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+
+val logger: Logger = LoggerFactory.getLogger("org.modelix.sample.restapimodelql.ModelServerLightWrapper")
+
+class LightModelClientWrapper {
+
+    private var MPS_MODEL_NAME: String = ""
+    private var WS_CONNECTION: String
+    lateinit var lightModelClient: LightModelClient
+
+    constructor(host: String = "localhost", port: Int = 48302, models: String) {
+        GeneratedLanguages.registerAll()
+        WS_CONNECTION = "ws://$host:$port/ws"
+        MPS_MODEL_NAME = models
+    }
+
+    suspend fun createConnection() {
+
+        // we require a http client with WS support for the connection
+        logger.info("Connecting to light model-server at $WS_CONNECTION")
+        val lightModelClient = LightModelClient(WebsocketConnection(HttpClient(CIO) { install(WebSockets) }, WS_CONNECTION))
+        this.lightModelClient = lightModelClient
+
+        // the modelQL query
+        this.lightModelClient.changeQuery(buildModelQuery {
+            // we traverse from the root (a repository)
+            root {
+                // to all modules
+                children("modules") {
+                    // selecting the ones that contain our desired model name
+                    whereProperty("name").contains(MPS_MODEL_NAME)
+                    children("models") {
+                        // and extract all root nodes, so Rooms and Courses, and their descendants
+                        children("rootNodes") {
+                            descendants { }
+                        }
+                    }
+                }
+            }
+        })
+        logger.info("Connection successful")
+    }
+
+    suspend fun getAllLectures(): List<N_Lecture> {
+        //TODO: can we avoid this cast by using generics in runAsyncRead?
+        logger.info("Loading all lectures")
+        return runAsyncRead(loadLectures) as List<N_Lecture>
+    }
+
+    suspend fun getAllRooms(): List<N_Room> {
+        logger.info("Loading all rooms")
+        return runAsyncRead(loadRooms) as List<N_Room>
+    }
+
+    private suspend fun runAsyncRead(givenFunction: (node: N_Repository) -> List<N_BaseConcept>?): List<N_BaseConcept>? {
+        var result: List<N_BaseConcept>? = null
+        kotlinx.coroutines.withTimeout(5.seconds) {
+            while (true) {
+                lightModelClient.checkException()
+                val node = lightModelClient.runRead { lightModelClient.getRootNode() }
+                if (node != null && lightModelClient.runRead { node.isValid }) {
+                    lightModelClient.runRead {
+                        logger.info("Reading repository")
+                        result = givenFunction(node.typed<N_Repository>())
+                    }
+                    break
+                }
+                delay(10.milliseconds)
```

Comment:
Why does the request for `node` need retrying in a loop?

Reply:
My interpretation was that we loop the checkException here, don/t we? (disclaimer: I got inspiration for this code from another repo)

Reply:
We loop as `node!=null` and `node.isValid` only is true if we actually get data from the remote.

Reply:
That's a piece of code that could need some explanations for readers of the sample to understand what they need to do. Or a better API that hides these things ;-)

Reply:
I will at some point in time yes. I added a TODO